### ISSUE TITLE
aws sam doesn’t create resource AWS::ApiGatewayV2::ApiMapping

### ISSUE THREAD

**Issue by makarono** [aws/serverless-application-model#2526]
### Description:
Sam skips AWS::ApiGatewayV2::ApiMapping on deployment into the cloud.
I added AWS::ApiGatewayV2::ApiMapping into SAM template, run sam package command , run sam deploy command . Sam reports "No changes to deploy. Stack api-test is up to date" . I tried several times with different api mapping and apis, sam simply doesn’t detects that resource. Debug log on deployment shows all resources which will be created/updated except that one.
When I put DependsOn in some resource just for test sam reports 
Error: Failed to create changeset for the stack: bla-bla, ex: Waiter ChangeSetCreateComplete failed: Waiter encountered a terminal failure state: For expression "Status" we matched expected path: "FAILED" Status: FAILED. Reason: Template format error: Unresolved resource dependencies [UserAPIMapping] in the Resources block of the template

    UserAPIMapping:
      Type: AWS::ApiGatewayV2::ApiMapping
      Properties:
        ApiId: !Ref ProgramsApi
        DomainName: !Ref DomainName
        ApiMappingKey: "user"
        Stage: !Ref ProgramsApi.Stage

### Steps to reproduce:
sam package
sam deploy

### Observed result:
Waiting for changeset to be created..
No changes to deploy. Stack httpapi-test is up to date

### Expected result:
expected result is that this gets deployed

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. os: System Version: macOS 12.6 (21G115)  Kernel Version: Darwin 21.6.0
3. If using SAM CLI, `sam --version`: SAM CLI, version 1.59.0
4. AWS region: eu-north-1

**Comment by aahung** [comment]
When you make change to the template, can you run `sam validate --debug` to dump the transformed template locally to see if `AWS::ApiGatewayV2::ApiMapping` in it?

Besides, did you run `sam build` after you made changes to your template? `sam package` might package the old built template.

**Comment by makarono** [comment]
@aahung 
I didn't run sam build because I have only one lambda with hello world and inline code. I deleted  old packaged template every time before running package command to avoid possibility that old file is not updated/overwritten en. 
For test only I created separate template with only  one resource AWS::ApiGatewayV2::ApiMapping and it was not detected by the sam when I was deploying to cloud.
Today I upgraded aws sam cli to latest version 1.60.0 before doing this test and I see that sam now detects and deploys the resource with same commands which I used first time `sam package` and `sam deploy`, thank you for the fix.   Here is the debug log


PACKAGE
=======

/tmp/test
❯❯❯ sam package --debug --template-file template-test.yaml --output-template-file packaged.yaml --s3-bucket test-bucket --profile default
2022-10-15 07:05:32,488 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2022-10-15 07:05:32,488 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2022-10-15 07:05:32,489 | Using config file: samconfig.toml, config environment: default
2022-10-15 07:05:32,489 | Expand command line arguments to:
2022-10-15 07:05:32,489 | --template_file=/private/tmp/test/template-test.yaml --output_template_file=packaged.yaml --s3_bucket=test-bucket
2022-10-15 07:05:32,517 | No Parameters detected in the template
2022-10-15 07:05:32,527 | There is no customer defined id or cdk path defined for resource HttpApiFunction, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,527 | There is no customer defined id or cdk path defined for resource Api, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,527 | There is no customer defined id or cdk path defined for resource APIMapping, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,527 | 0 stacks found in the template
2022-10-15 07:05:32,726 | There is no customer defined id or cdk path defined for resource HttpApiFunction, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,727 | There is no customer defined id or cdk path defined for resource Api, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,727 | There is no customer defined id or cdk path defined for resource APIMapping, so we will use the resource logical id as the resource id
2022-10-15 07:05:32,727 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:05:32,727 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:05:32,727 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping

Successfully packaged artifacts and wrote output template to file packaged.yaml.
Execute the following command to deploy the packaged template
sam deploy --template-file /private/tmp/test/packaged.yaml --stack-name <YOUR STACK NAME>


VALIDATE
=======

/tmp/test
❯❯❯ sam validate -t packaged.yaml --debug
2022-10-15 07:05:38,088 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2022-10-15 07:05:38,088 | Using config file: samconfig.toml, config environment: default
2022-10-15 07:05:38,088 | Expand command line arguments to:
2022-10-15 07:05:38,088 | --template_file=/private/tmp/test/packaged.yaml
2022-10-15 07:05:38 Loading policies from IAM...
2022-10-15 07:05:42 Finished loading policies from IAM.
2022-10-15 07:05:42,602 | Translated template is:
AWSTemplateFormatVersion: '2010-09-09'
Description: api
Resources:
  APIMapping:
    Type: AWS::ApiGatewayV2::ApiMapping
    Properties:
      ApiId:
        Ref: Api
      DomainName: my.domain.com
      ApiMappingKey: user
      Stage:
        Ref: Apiv1Stage
    Metadata:
      SamResourceId: APIMapping
  HttpApiFunction:
    Metadata:
      SamResourceId: HttpApiFunction
    Properties:
      Code:
        ZipFile: "def handler(event, context):\n    import json\n    return {\n  \
          \      \"statusCode\": 200,\n        \"body\": json.dumps(event),\n    }\n"
      Handler: index.handler
      Role:
        Fn::GetAtt:
        - HttpApiFunctionRole
        - Arn
      Runtime: python3.9
      Tags:
      - Key: lambda:createdBy
        Value: SAM
    Type: AWS::Lambda::Function
  HttpApiFunctionRole:
    Properties:
      AssumeRolePolicyDocument:
        Statement:
        - Action:
          - sts:AssumeRole
          Effect: Allow
          Principal:
            Service:
            - lambda.amazonaws.com
        Version: '2012-10-17'
      ManagedPolicyArns:
      - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
      Tags:
      - Key: lambda:createdBy
        Value: SAM
    Type: AWS::IAM::Role
  HttpApiFunctionExplicitApiPermission:
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Ref: HttpApiFunction
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Sub:
        - arn:${AWS::Partition}:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/GET/path
        - __ApiId__:
            Ref: Api
          __Stage__: '*'
    Type: AWS::Lambda::Permission
  Api:
    Metadata:
      SamResourceId: Api
    Properties:
      Body:
        info:
          version: '1.0'
          title:
            Ref: AWS::StackName
        paths:
          /path:
            get:
              x-amazon-apigateway-integration:
                payloadFormatVersion: '1.0'
                type: aws_proxy
                uri:
                  Fn::Sub: arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${HttpApiFunction.Arn}/invocations
                httpMethod: POST
                timeoutInMillis: 30000
              security:
              - AWS_IAM: []
              responses: {}
        x-amazon-apigateway-cors:
          allowHeaders:
          - '*'
          allowMethods:
          - OPTIONS
          - POST
          - GET
          allowOrigins:
          - '*'
          maxAge: 600
        tags:
        - name: httpapi:createdBy
          x-amazon-apigateway-tag-value: SAM
        openapi: 3.0.1
        components:
          securitySchemes:
            AWS_IAM:
              in: header
              name: Authorization
              type: apiKey
              x-amazon-apigateway-authtype: awsSigv4
      FailOnWarnings: true
    Type: AWS::ApiGatewayV2::Api
  Apiv1Stage:
    Properties:
      ApiId:
        Ref: Api
      AutoDeploy: true
      DefaultRouteSettings:
        ThrottlingBurstLimit: 200
      RouteSettings:
        GET /path:
          ThrottlingBurstLimit: 600
      StageName: v1
      StageVariables:
        StageVar: Value
      Tags:
        httpapi:createdBy: SAM
    Type: AWS::ApiGatewayV2::Stage

/private/tmp/test/packaged.yaml is a valid SAM Template



DEPLOY
=======

/tmp/test
✖ sam deploy --debug --no-fail-on-empty-changeset --template-file packaged.yaml --parameter-overrides ParameterKey=EnvType,ParameterValue=stage --tags Enviroment=stage --stack-name httpapi-test --s3-bucket test-bucket --capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM CAPABILITY_AUTO_EXPAND
2022-10-15 07:08:00,913 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2022-10-15 07:08:00,913 | Using config file: samconfig.toml, config environment: default
2022-10-15 07:08:00,913 | Expand command line arguments to:
2022-10-15 07:08:00,913 | --template_file=/private/tmp/test/packaged.yaml --parameter_overrides={'EnvType': 'stage'} --tags={'Enviroment': 'stage'} --stack_name=httpapi-test --s3_bucket=test-bucket --capabilities=('CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND') --on_failure=ROLLBACK
2022-10-15 07:08:01,057 | No Parameters detected in the template
2022-10-15 07:08:01,067 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:08:01,067 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:08:01,067 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping
2022-10-15 07:08:01,068 | 0 stacks found in the template
2022-10-15 07:08:01,387 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:08:01,388 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:08:01,388 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping
2022-10-15 07:08:01,388 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:08:01,388 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:08:01,388 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping

	Deploying with following values
	===============================
	Stack name                   : httpapi-test
	Region                       : eu-north-1
	Confirm changeset            : False
	Disable rollback             : False
	Deployment s3 bucket         : test-bucket
	Capabilities                 : ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
	Parameter overrides          : {"EnvType": "stage"}
	Signing Profiles             : {}

Initiating deployment
=====================
2022-10-15 07:08:01,459 | No Parameters detected in the template
2022-10-15 07:08:01,469 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:08:01,469 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:08:01,469 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping
2022-10-15 07:08:01,469 | 0 stacks found in the template
2022-10-15 07:08:01,469 | No Parameters detected in the template
2022-10-15 07:08:01,477 | Sam customer defined id is more priority than other IDs. Customer defined id for resource HttpApiFunction is HttpApiFunction
2022-10-15 07:08:01,477 | Sam customer defined id is more priority than other IDs. Customer defined id for resource Api is Api
2022-10-15 07:08:01,477 | Sam customer defined id is more priority than other IDs. Customer defined id for resource APIMapping is APIMapping
2022-10-15 07:08:01,478 | 3 resources found in the stack
2022-10-15 07:08:01,478 | Found Serverless function with name='HttpApiFunction' and InlineCode
2022-10-15 07:08:01,478 | Detected Inline Swagger definition
2022-10-15 07:08:01,478 | Auth checks done on swagger are not exhaustive!
2022-10-15 07:08:01,745 | Stack with id httpapi-test does not exist
Uploading to 8c1b56a623b97b772f1edbfed1add237.template  1597 / 1597  (100.00%)

Waiting for changeset to be created..
CloudFormation stack changeset
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Operation                                   LogicalResourceId                           ResourceType                                Replacement
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ Add                                       APIMapping                                  AWS::ApiGatewayV2::ApiMapping               N/A
+ Add                                       Apiv1Stage                                  AWS::ApiGatewayV2::Stage                    N/A
+ Add                                       Api                                         AWS::ApiGatewayV2::Api                      N/A
+ Add                                       HttpApiFunctionExplicitApiPermission        AWS::Lambda::Permission                     N/A
+ Add                                       HttpApiFunctionRole                         AWS::IAM::Role                              N/A
+ Add                                       HttpApiFunction                             AWS::Lambda::Function                       N/A
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Changeset created successfully. arn:aws:cloudformation:eu-north-1:967256687856:changeSet/samcli-deploy1665810481/3cd2909b-c1e8-413b-b49d-cdcf421457ff


2022-10-15 07:08:08 - Waiting for stack create/update to complete

CloudFormation events from stack operations (refresh every 0.5 seconds)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ResourceStatus                              ResourceType                                LogicalResourceId                           ResourceStatusReason
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
CREATE_IN_PROGRESS                          AWS::CloudFormation::Stack                  httpapi-test                                User Initiated
CREATE_IN_PROGRESS                          AWS::IAM::Role                              HttpApiFunctionRole                         -
CREATE_IN_PROGRESS                          AWS::IAM::Role                              HttpApiFunctionRole                         Resource creation Initiated
CREATE_COMPLETE                             AWS::IAM::Role                              HttpApiFunctionRole                         -
CREATE_IN_PROGRESS                          AWS::Lambda::Function                       HttpApiFunction                             -
CREATE_IN_PROGRESS                          AWS::Lambda::Function                       HttpApiFunction                             Resource creation Initiated
CREATE_COMPLETE                             AWS::Lambda::Function                       HttpApiFunction                             -
CREATE_IN_PROGRESS                          AWS::ApiGatewayV2::Api                      Api                                         -
CREATE_IN_PROGRESS                          AWS::ApiGatewayV2::Api                      Api                                         Resource creation Initiated
CREATE_COMPLETE                             AWS::ApiGatewayV2::Api                      Api                                         -
CREATE_IN_PROGRESS                          AWS::ApiGatewayV2::Stage                    Apiv1Stage                                  -
CREATE_IN_PROGRESS                          AWS::Lambda::Permission                     HttpApiFunctionExplicitApiPermission        -
CREATE_IN_PROGRESS                          AWS::Lambda::Permission                     HttpApiFunctionExplicitApiPermission        Resource creation Initiated
CREATE_IN_PROGRESS                          AWS::ApiGatewayV2::Stage                    Apiv1Stage                                  Resource creation Initiated
CREATE_COMPLETE                             AWS::ApiGatewayV2::Stage                    Apiv1Stage                                  -
CREATE_IN_PROGRESS                          AWS::ApiGatewayV2::ApiMapping               APIMapping                                  -
CREATE_FAILED                               AWS::ApiGatewayV2::ApiMapping               APIMapping                                  Invalid domain name identifier specified
                                                                                                                                    (Service: AmazonApiGatewayV2; Status
                                                                                                                                    Code: 404; Error Code: NotFoundException;
                                                                                                                                    Request ID:
                                                                                                                                    427ddf6a-ac5b-4cf6-80a1-c57d40fd27b1;
                                                                                                                                    Proxy: null)
CREATE_FAILED                               AWS::Lambda::Permission                     HttpApiFunctionExplicitApiPermission        Resource creation cancelled
ROLLBACK_IN_PROGRESS                        AWS::CloudFormation::Stack                  httpapi-test                                The following resource(s) failed to
                                                                                                                                    create:
                                                                                                                                    [HttpApiFunctionExplicitApiPermission,
                                                                                                                                    APIMapping]. Rollback requested by user.

**Comment by makarono** [comment]
@aahung I didn't run `sam build` because  there was only one "Hello world" function in template with inline code. I figured out what is the problem. There was an indentation problem in my sam yaml template. My AWS::ApiGatewayV2::ApiMapping was indented too much so it look like "AWS::ApiGatewayV2::ApiMapping" is part of "AWS::Serverless::HttpApi". cfn-lint command didn't detect any errors in template and my yaml linter also didn't detect errors because yaml was valid. `sam package` command simply removed wrongly indented "AWS::ApiGatewayV2::ApiMapping" resource from packaged template so resource was not deployed.
When I fixed indentation everything  was deployed correctly. So error was on my side and I apologise for raised issue.
My problem wouldn’t exist if cnf-lint linter would detect incorrect code bloks which are not part of "AWS::Serverless::HttpApi" and if `sam package` command wouldn’t just silently delete incorrectly defined code.
